backend/controllers/prepostAppoinments.js

```diff
@@ -0,0 +1,226 @@
+          lastName,
+          offenderNo,
+          appointmentType: appointmentTypeDescription,
+          location: locationDescription,
+          startTime,
+          endTime,
+          comment,
+          recurring,
+          times,
+          repeats,
+        }),
+      })
+    } catch (error) {
+      logError(req.originalUrl, error, serviceUnavailableMessage)
+      res.render('error.njk')
+    }
+  }
+  const createAppointment = async (context, appointmentDetails) => {
+    const {
+      startTime,
+      endTime,
+      comment,
+      recurring,
+      times,
+      repeats,
+      bookingId,
+      locationId,
+      appointmentType,
+    } = appointmentDetails
+
+    const mainAppointment = {
+      appointmentDefaults: {
+        comment,
+        locationId: Number(locationId),
+        appointmentType,
+        startTime,
+        endTime,
+      },
+      appointments: [{ bookingId }],
+      repeat:
+        recurring === 'yes'
+          ? {
+              repeatPeriod: repeats,
+              count: times,
+            }
+          : undefined,
+    }
+
+    await elite2Api.addAppointments(context, mainAppointment)
+  }
+
+  const post = async (req, res) => {
+    const { offenderNo } = req.params
+    const {
+      postAppointment,
+      preAppointment,
+      postAppointmentDuration,
+      preAppointmentDuration,
+      preAppointmentLocation,
+      postAppointmentLocation,
+    } = req.body
+
+    try {
+      const appointmentDetails = unpackAppointmentDetails(req)
+      const {
+        startTime,
+        endTime,
+        comment,
+        recurring,
+        times,
+        repeats,
+        locationDescription,
+        appointmentTypeDescription,
+        locationTypes,
+        firstName,
+        lastName,
+      } = appointmentDetails
+
+      packAppointmentDetails(req, appointmentDetails)
+
+      const errors = validate({ preAppointment, postAppointment, preAppointmentLocation, postAppointmentLocation })
+
+      if (errors.length) {
+        return res.render('prepostAppointments.njk', {
+          links: getLinks(offenderNo),
+          locations: locationTypes,
+          formValues: {
+            postAppointment,
+            preAppointment,
+            postAppointmentDuration,
+            preAppointmentDuration,
+            preAppointmentLocation: preAppointmentLocation && Number(preAppointmentLocation),
+            postAppointmentLocation: postAppointmentLocation && Number(postAppointmentLocation),
+          },
+          errors,
+          details: toAppointmentDetailsSummary({
+            firstName,
+            lastName,
+            offenderNo,
+            appointmentType: appointmentTypeDescription,
+            location: locationDescription,
+            startTime,
+            endTime,
+            comment,
+            recurring,
+            times,
+            repeats,
+          }),
+        })
+      }
+
+      await createAppointment(res.locals, appointmentDetails)
```

Comment:
These don't fail together, we may need to make use of the bulk appointments API, however a special role is required to use it. We should revisit this later.